draft interfaces for replication admission control v2

sumeerbhola opened the pull request:
This is extremely rough, and will likely be hard to understand. 
Meant for folks working on replication AC v2 prototype.

sumeerbhola committed:
draft interfaces for replication admission control v2

Commit:
roachtest: add perturbation tests

This commit adds a set of roachtests and a framework to measure the
impact to latency of various perturbations to the system.

The roachtest names all have the format
perturbation/[full|dev|metamorphic]/[perturbationType]

The four perturbationType variations added in this initial commit are:
restart
partition
add-node
decommission

Epic: CRDB-37514

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124484)
<!-- Reviewable:end -->

sumeerbhola committed:
kvflowcontroller: modify bucket implementation to support waiting for quorum

The existing implementation allowed for sequentially waiting on all
buckets (btw, these should not be called bucket, since this is not a
token bucket). We now need to wait for a quroum in some cases.

A combination of bucket.TokensAvailable and WaitForHandlesAndChannels
is used for such waiting.

sumeerbhola committed:
cleanup the store streams interfaces and move them into a separate file

kvoli committed:
kvflowcontrol: [dnm] add basic store send stream token watcher impl

Introduce a basic implementation of the StoreStreamTokenWatcher
interface. The implementation will notify callers when there are tokens
available for a `WorkClass,SendTokenCounter` pair. The caller should
then attempt deducting tokens.

Epic: none
Release note: None

sumeerbhola committed:
rough code for RangeController.{WaitForEval,HandleRaftEvent}

sumeerbhola committed:
copy over Tracker and tweak it

sumeerbhola committed:
replicaSendStream, some code

sumeerbhola committed:
replicaSendStream.advanceNextRaftIndexAndQueued etc.

sumeerbhola committed:
replicaSendStream, some code

sumeerbhola committed:
RangeControllerImpl, partial code

sumeerbhola committed:
unify SentTokenCounter and EvalTokenCounter

sumeerbhola committed:
copied bucket to make tokenCounter

WaitForEval is the new waiting interface.

kvoli committed:
kvflowconnectedstream: [dnm] clean up and comment send token watcher

Clean up the store send stream token watcher implementation and add
relevant comments to exported functions.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] update token watcher to use updated handle

The handle token counter interface was updated in the prior commit to
add `TryDeductAndUnblockNextWaiter`. Call this method when the handle
wait channel is signaled, before proceeding to notify the respective
token watchers.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] impl token counter and handle

Add implementations for the `TokenCounter` and `TokenWaitingHandle`
methods, using `tokenCounter`.

Release note: None
Epic: none

sumeerbhola committed:
RangeControllerImpl.WaitForEval

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+// The Ready described here is only the subset that is needed by replication
+// AC -- heartbeats, appending to the local log etc. are not relevant.
+//
+// Ready must be called on every tick/ready of Raft since we cannot tolerate
+// state transitions from
+//
+// StateReplicate => {state in all states: state != StateReplicate} =>
+// StateReplicate
```
Are you saying that we can't tolerate an ABA problem in which we go out and back into `StateReplicate` and can't tell if the node has always been in this state or not?

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+// Ready must be called on every tick/ready of Raft since we cannot tolerate
+// state transitions from
+//
+// StateReplicate => {state in all states: state != StateReplicate} =>
```
Is it "any" rather than "all"?

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+// that is not observed by RangeController. Since outgoing messages are sent
+// in the tick/ready handling and a return from the intermediate state to
+// StateReplicate relies on receiving responses to those outgoing messages, we
+// should not miss any transitions. If stale messages received in a step can
+// cause transitions out and back without observation, we can add a monotonic
+// counter for each follower inside Raft (this is just local state at the
+// leader), which will be incremented on every state transition and expose
+// that via the RaftInterface.
```
One such counter is the `Term` of the raft instance. But it still doesn't cover cases when one term goes out/in `StateReplicate`.

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+// internal mutexes in RangeController.
+type RaftInterface interface {
+	// FollowerState returns the current state of a follower. The value of
+	// nextUpperBound is populated iff in StateReplicate. All entries >=
+	// nextUpperBound have not yet had MsgApps constructed.
```
In case it impacts correctness reasoning somewhere, "have not yet had MsgApps constructed" has an implicit asterisk: "by this connected stream / StateReplicate" (at some concrete in-memory "epoch" if we number them).

There can be in-flight `MsgApp`s at higher indices, and we may even get acks for them (beyond the `Next` index). E.g. if we just transitioned out / back in `StateReplicate` and lost some tracking information.

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+	// When a follower transitions from {StateProbe,StateSnapshot} =>
+	// StateReplicate, or was already in StateReplicate but was disconnected and
+	// has now reconnected, we can now start trying to send MsgApps. We should
+	// notice such transitions both in HandleRaftEvent and SetReplicas. We
+	// should *not* construct MsgApps for a StateReplicate follower that is
+	// disconnected -- there is no timeliness guarantee on how long a follower
+	// will stay in StateReplicate despite it being down, and by sending such a
+	// follower MsgApps that are not being received we are defeating flow
+	// control (since we will have advanced nextUpperBound).
```
Do we intend to stay in `StateReplicate` if the follower is disconnected? Today, there is a `ReportUnreachable` call that puts the flow to `StateProbe`. If this method is not called, raft will be sending `MsgApps` until it reaches max-inflight limits and stalls (there won't be more than one probe because a disconnected follower does not reply to heartbeats).

I'm guessing we're moving control over "raft will be sending `MsgApps` until it reaches max-inflight limits and stalls" to AC. So we may choose not to send any `MsgApps`, not call `ReportUnreachable`, and stay in `StateReplicate`. However, when we reconnect, there is a non-zero chance of a reject and needing to probe.

But I guess we won't need any special handling for this disconnected state because we will unmount the "connected stream", and won't be sending any messages as a result.

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+	FollowerTransportConnected(storeID roachpb.StoreID) bool
+	// HighestEntryIndex is the highest index assigned in the log, and produced
+	// in Ready.Entries(). If there have been no entries produced, since this
+	// replica became the leader, this is the commit index.
```
A newly elected leader can have a suffix of entries which are not yet committed. Should we initialize to the log size, rather than to the commit index?

A thing about the commit index on the new leader: it doesn't know it before it commits a dummy entry at its term. So, by the time the leader learns the commit index, it has already appended an entry to its log.

Do we need to tie this definition to `Ready.Entries()`? We can get the size of the log more directly, by accessing `raft.raftLog.lastEntryID()`. Would it suffice to say that this returns the last index in the log at all times, and not mention the "becoming leader" case?

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+	//
+	// If the sum of all entries in [start,end) are <= maxSize, all will be
+	// returned. Else, entries will be returned until, and including, the first
+	// entry that causes maxSize to be equaled or exceeded. This implies at
+	// least one entry will be returned in the MsgApp on success.
```
Today, various raft log internals return a prefix exceeding the `maxSize` only if it's a single entry; otherwise it's always `<= maxSize`, and the offender entry gets dropped from the response.

On a few occasions, I considered this logic unfortunate (we waste a read) and corner-casey, and contemplated to simplify it into the one you described: allow exceeding the size always.

Is this nuance critical here, or is an implementation detail?

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go:
```diff
@@ -0,0 +1,1339 @@
+	// least one entry will be returned in the MsgApp on success.
+	//
+	// Returns raft.ErrCompacted error if log truncated. If no error,
+	// nextUpperBound is advanced to be equal to end. If raft.ErrCompacted is
```
It's probably advanced to the index following the last returned entry, which is not always `end` if `maxSize` is exceeded.

sumeerbhola committed:
more RangeControllerImpl

sumeerbhola committed:
more replicaSendStream

kvoli committed:
kvflowconnectedstream: [dnm] further test store stream watcher

Add some additional testing for Updating a handle in
`StoreStreamSendTokensWatcher`. Add setting `onChanged` functionality for
the TokenCounter, copied mostly from the kvflowcontrol controller.

Epic: none
Release note: None

sumeerbhola committed:
replicaSendStream, tracker

pav-kv committed:
raft: support lazy appends

Ports PR 125002

kvoli committed:
kvflowconnectedstream: [dnm] implement wait for eval

Implement the `WaitForEval` functionality, which waits the supplied
token wait handles until (a) all required handles are signaled+available
and (b) at least the required quorum of handles is signaled+available.

Note checking for availability is done after satisfying (a)+(b) for
signaling alone. See the TODO for a potential hazard here and trade-off
we must decide upon.

Epic: none
Release note: None

sumeerbhola committed:
started fixing commentary in stream.go

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124484)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 205 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618512397:-Nz8EavaAFL_1GcgFhO-:bo1e6en) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L205)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Is it "any" rather than "all"?
</blockquote></details>

I think we are saying the same thing -- it was meant as the "for all", "for any" universal quantification, though it was probably a misuse in expressing "set comprehension". Removed now.
___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 206 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618508219:-Nz8FcX5CkNpB7KLgBuK:bj6x34g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L206)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Are you saying that we can't tolerate an ABA problem in which we go out and back into `StateReplicate` and can't tell if the node has always been in this state or not?
</blockquote></details>

I've dropped all this language for a few reasons:
- We want to keep a replicaSendStream briefly around in `StateProbe`, given that a single nack causes a transition to `StateProbe`. So we need to be able to tolerate a regression in `indexToSend`.
- You brought up the old ack case ("There can be in-flight `MsgApp`s at higher indices, and we may even get acks for them"), which means `indexToSend` can advance while in StateReplicate.
- This requirement of observing all state transitions is too strong, and is unnecessary given we need to handle the previous bullets.

So the code should just react to the state that we are currently in, and make any adjustments needed based on the current (Match, Next).
___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 215 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618517358:-Nz8GbJw48Qlm5WCjYyc:bmc2sim) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L215)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

One such counter is the `Term` of the raft instance. But it still doesn't cover cases when one term goes out/in `StateReplicate`.
</blockquote></details>

Dropped this part.
___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 249 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618528467:-Nz8e4wc3_R372ojO6sh:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L249)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

In case it impacts correctness reasoning somewhere, "have not yet had MsgApps constructed" has an implicit asterisk: "by this connected stream / StateReplicate" (at some concrete in-memory "epoch" if we number them).

There can be in-flight `MsgApp`s at higher indices, and we may even get acks for them (beyond the `Next` index). E.g. if we just transitioned out / back in `StateReplicate` and lost some tracking information.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
added a comment on robust token returns

sumeerbhola committed:
design for robust flow token return in RACv2

sumeerbhola committed:
interface tweaks in response to Pavel's PR comments

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124484)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 259 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618567528:-NzFb0b73XY6latzObeA:bl3ivco) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L259)):*
> Do we intend to stay in `StateReplicate` if the follower is disconnected? Today, there is a `ReportUnreachable` call that puts the flow to `StateProbe`.

We will keep that functionality. In that either the RaftTransport queue will overflow, or the circuit breaker will trip (as discussed in https://cockroachlabs.slack.com/archives/C06UFBJ743F/p1716468360084809) even if nothing is being sent, and the transition to StateProbe will happen. At that point we will not be sending any more MsgApps even if there are flow tokens available because the `replicaSendStream` will be closed.

The code below needs some updating to reflect this reality. Specifically, the `replicateSoftDisconnected` is unnecessary, and I will remove it.
___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 274 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618589282:-NzFehtB1dVWTnWC3k3k:b-jahmnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L274)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

A newly elected leader can have a suffix of entries which are not yet committed. Should we initialize to the log size, rather than to the commit index?

A thing about the commit index on the new leader: it doesn't know it before it commits a dummy entry at its term. So, by the time the leader learns the commit index, it has already appended an entry to its log.

Do we need to tie this definition to `Ready.Entries()`? We can get the size of the log more directly, by accessing `raft.raftLog.lastEntryID()`. Would it suffice to say that this returns the last index in the log at all times, and not mention the "becoming leader" case?
</blockquote></details>

Good point. Changed to

```
	// LastEntryIndex is the highest index assigned in the log.
	LastEntryIndex() uint64
```
___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 283 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618603976:-NzFexuYCGsREqKvJBcX:bb5e3ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L283)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Today, various raft log internals return a prefix exceeding the `maxSize` only if it's a single entry; otherwise it's always `<= maxSize`, and the offender entry gets dropped from the response.

On a few occasions, I considered this logic unfortunate (we waste a read) and corner-casey, and contemplated to simplify it into the one you described: allow exceeding the size always.

Is this nuance critical here, or is an implementation detail?
</blockquote></details>

I think the nuance is important here, in that for throttled stores, the tokens are precious and we'd rather go over slightly, rather than under utilize the tokens a `replicaSendStream` has grabbed and stashed in `sendQueue.deductedForScheduler`. It will return them, if it underutilizes, but there a delay incurred until it returns,  when no one is using these available tokens.

Your point about wasting a read makes me even more convinced that we should do this.
___
*[`pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go` line 286 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/124484#gh-1618607784:-NzFgp9r85-RTp1HPdkY:b-8wc6yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e3632f73f50265bfda802a532b3629851c1fb/pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go#L286)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

It's probably advanced to the index following the last returned entry, which is not always `end` if `maxSize` is exceeded.
</blockquote></details>

Ah yes. This was a mistake -- thanks. Fixed.


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
section on correctness with priority override etc.

sumeerbhola committed:
more commentary on priority inheritance

sumeerbhola committed:
new Tracker interface with implementation todos

kvoli committed:
kvflowconnectedstream: [dnm] update tracker methods

Update the tracker methods to use a RaftPriority, instead of the
prior admission work priority and work class.

Note that callers were not updated to map existing work priorities /
classes into a RaftPriority and instead just cast. This is a TODO.

Epic: none
Release note: None

sumeerbhola committed:
lot of cleanup in stream.go

pav-kv committed:
Add basic RaftInterface implementation

sumeerbhola committed:
more replicaSendStream

sumeerbhola committed:
improved text for Admitted behavior in Raft and RAC. RaftAdmittedInterface for
raft behavior at all replicas related to admitted advancement.

sumeerbhola committed:
some more force-flush todos

kvoli commented on pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/token_counter.go:
Chatting w/ @sumeerbhola:

We should unblock the next waiter immediately. Utilize the normal back-pressure the goroutine scheduling is giving us. 

Refresh is only being closed when the replica set changes. Otherwise, we expect this to successfully wait or get cancelled.

Worst case burst of GOMAXPROCS acquiring tokens when signaled. 

TODO(kvoli): update code to try deducting immediately after being signaled on a handle's channel

kvoli committed:
kvflowconnectedstream: [dnm] update wait for eval to deduct

WaitForEval would collect signals from the waiting handles, then after
having been signaled on a quorum and all required handles, would try
deducting tokens.

This was problematic, since it blocked other waiters on the same store
send stream. Now we try to deduct and signal immediately following a
signal.

Epic: none
Release note: None

sumeerbhola committed:
- move raft impl to raft.go
- remove Ready.RetransmitMsgApps()

sumeerbhola committed:
integration: construction/destruction of RangeController on Replica

sumeerbhola committed:
integration: make AdmitKVWork call RangeController.WaitForEval

sumeerbhola committed:
integration: plumbing for priority encoding in entry and message sending

sumeerbhola committed:
more comments in stream.go

sumeerbhola committed:
integration: code scaffolding for advancing admitted

sumeerbhola committed:
integration: undo my stupidity around RaftPriority and the unused zero value

sumeerbhola committed:
integration: more code

sumeerbhola committed:
kvflowtokentracker: undo comment change

sumeerbhola committed:
integration: description comment

kvoli committed:
kvflowconnectedstream: [dnm] add range controller dd test

Add a data-driven test for the `RangeController` implementation. A basic
testcase is added in `range_controller_basic`, which exercises sending
and admitting entries on single range fixed replica set.

The commands available are:

```
init range_id=<range_id> tenant_id=<tenant_id> local_replica_id=<local_replica_id>
store_id=<store_id> replica_id=<replica_id> type=<type>
...

set_replicas
store_id=<store_id> replica_id=<replica_id> type=<type>
...

send
range_id=<range_id> size=<size> pri=<pri>
...

admit
store_id=<store_id> to=<to> pri=<pri>
...
```

Epic: none
Release note: None

sumeerbhola committed:
integration: waitingForAdmissionState implementation, and commentary

kvoli committed:
kvflowconnectedstream: [dnm] fix eval/send token counter aliasing

Introduce separate Eval/Send token counters. These were backed by the
same token counter previously, leading to incoherent token deduction.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] include send queue in dd test output

Include send queue information, if one exists, when printing out the
state in `TestRangeController`. For example:

A queue exists for entry 4, of size 1024 KiB at a normal priority.

```
... queue=[4,5) size=+1021 KiB pri=NormalPri
```

Epic: none
Release note: None

sumeerbhola committed:
minor comment cleanup

kvoli committed:
kvflowconnectedstream: [dnm] wait for async notify in dd test

The `Notify()` callback is scheduled async, which is triggered from
returning tokens via `admit`. Wait a short duration to ensure these
callbacks process, before returning from `admit`.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] add dd replica changes test

Incomplete, see todos for known bugs.

Epic: none
Release note: None

sumeerbhola committed:
integration: processing of piggybacked MsgAppResp

kvoli committed:
kvflowconnectedstream: [dnm] update impl todo commentary

Update misc comments about remaining TODOs and bugs to squash.

Epic: none
Release note: None

pav-kv committed:
draft impl for RaftAdmittedInterface

kvoli committed:
kvflowconnectedstream: [dnm] fix bug to actually set replicas

`SetReplicas` would not update the range controller replica set, just
the replica map, which contains the replica state. Update the replica
set as well.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] add multi-range support and testing

Add multi-range support for `TestRangeController`, adding a new test
`range_controller_multi_range`. The syntax is updated to support
specifying commands for a specific range.

Epic: none
Release note: None

pav-kv committed:
raft: track the stable log index

kvoli committed:
.*: update todo assignment

Epic: none
Release note: None

pav-kv committed:
kvserver: enable raft lazy appends by default

kvoli committed:
kvserver: [dnm] add sync for rac_v2 integration

Add synchronization around replica integration, which is used to mediate
access to other datastructures beneath it, such as the range controller.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] make range controller test deterministic

Previously the range output would print ranges in a non-deterministic
order. Sort the keys before iterating over ranges to print, in order to
make the output deterministic.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: [dnm] test replica tracker state changes

Update the range controller test to allow changing a replicas tracker
state between StateReplicate, StateProbe and StateSnapshot. The replica
tracker state is then updated in the range_controller_replica_changes
test.

Epic: none
Release note: None

kvoli committed:
.*: move raft prio to kvflowcontrolpb

Also update the encoding/decoding methods to use the implemented methods
in the raftlog. Note the encoding is currently broken, see the comment
left in `pkg/kvkvserver/raftlog/payload.go`.

Epic: none
Release note: None

kvoli committed:
raftlog/kvflowcontrol: fix priority decoding bug and test

The raft priority was not being decoded to account for the 6 bit left
shift. Shift the bits right 6 to decode the priority correctly.

Epic: none
Release note: None

kvoli committed:
kvserver: add basic rac v2 integration test

Add `TestRACV2Basic` which starts a 3 node cluster and issues a single
put against the scratch range. Currently, the test fails to finish
starting up the nodes in the cluster, stalling.

Epic: none
Release note: None

kvoli committed:
kvserver: add 3 node variation and fix misc panics

The 3 node variety still panics, the 1 node variety now runs.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: add basic wait for eval sync

Wrap fields accessed by WaitForEval in the RangeControllerImpl in a
mutex to synchronize access.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: advance next raft index on non-ac entries

Entries which weren't subject to rac v2 would not advance the next index
tracking state. Fix this.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: sync connected stream

Epic: none
Release note: None

pav-kv committed:
kvserver: test priorityInheritanceState

Epic: none
Release note: none

sumeerbhola committed:
add raftMu.AssertHeld in various integration methods

kvoli committed:
kvflowconnectedstream: sprinkle more log statements

Introduce more v(1) log statements on entry processing.

Epic: none
Release note: None

pav-kv committed:
kvserver: straighten up priorityInheritanceState

Epic: none
Release note: none

sumeerbhola committed:
fiddling with starting state of send queue

kvoli committed:
raft/kvserver: set msg.from and log when sending messages

Fix two bugs: (1) uninitialized map in the integration struct and (2)
the replica sending a message would not populate the `From` field.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: fix negative queue size bug

When calculating the send queue size, two unsigned integers were
deducted then checked if the result was less than 0. This is impossible
as the result is also unsigned. Update to use signed integers.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: use string locked on panic in locked method

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: elaborate panics / logging

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: increase panic verbosity

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: allow leader next > index_to_send

A prior commit enforced the invariant that the leader's next will
advance prior to the range controller processing the corresponding
entries.

Account for this and relax the invariant that next <= index_to_send to
only apply to follower replicas.

Epic: none
Release note: None

sumeerbhola committed:
changes to invariant checking for leader

kvoli committed:
kvflowconnectedstream: ensure send token handles are created/cleaned up

It was possible for a send stream to be initialized for a follower with
a pre-existing queue. The queue would then never be processed as a send
token wait handle was not created. Ensure a handle is created and
subsequently cancelled when the send queue is empty.

Epic: none
Release note: None

pav-kv committed:
raft: add NextUnstableIndex method

Epic: none
Release note: none

kvoli committed:
kvflowconnectedstream: fix range controller test

Epic: none
Release note: None

sumeerbhola committed:
set leader for single node case

kvoli committed:
kvserver: add rac v2 lease transfer test

Add `TestRACV2Basic/lease_transfer`, which creates a 3 node cluster then
transfers the lease from s1 to s3 and writes some data.

This test currently fails post-transfer with:

```
entry_index=25 != index_to_send=26
```

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: init send stream to next unstable index

When initializing a new replica send stream, either on descriptor change
or a raft event, use the next unstable index to determine the initial
next raft index.

Epic: none
Release note: None

kvoli committed:
kvserver: add rac v2 basic relocate range test

Passes 1k/1k runs, with tweaks in this commit.

Epic: none
Release note: None

sumeerbhola committed:
cleanup nextRaftIndex initialization for new streams; add missing call to kvflowcontrolpb.AdmissionPriorityToRaftPriority

kvoli committed:
kvserver: assert raft pri bounds and add rac v2 limit test

The additional test limits the tokens to 4 MiB, from the default 16 MiB
and then sends 500 x 512 KiB writes (250 MiB).

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: sync store streams token counter

Epic: none
Release note: None

kvoli committed:
kvserver: update rac v2 basic integration test params

Epic: none
Release note: None

kvoli committed:
raft,kvserver: advance progress admitted on leader

When receiving a message which advances the admitted index for a
follower, reflect this in the progress tracker.

Epic: none
Release note: None

kvoli committed:
kvserver: fix racv2 basic test for relocate

kvoli committed:
kvflowcontrolpb: assert valid raft pri after conversion

Asserting on a valid raft priority value was done before converting,
which could lead to false invariant violations. Move checking to later.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: fix negative queue size

Only deduct tokens from the queue size sum when they are >= next raft
initial index. Otherwise, tokens are deducted on sending for entries the
queue never incremented in.

Epic: none
Release note: None

sumeerbhola committed:
small assertion change in RaftPriorityConversionForUnusedZero

kvoli committed:
kvserver: thread ctx through rac v2 methods

Thread `ctx` through several rac v2 methods so that log statements
include the range information and other ambient log tags for debugging.

Epic: none
Release note: None

kvoli committed:
kvserver: replace panic w/ fatal and fix misc bugs

Update existing assertions to use `log.Fatalf` with an assertion failed
error, instead of `panic`, in order to include the context tags on crash
reporting.

The `storeStreamSendTokensWatcher` was incorrectly updating watcher
handles, resulting in trackiing zombie handle references which would
panic when dereferenced.

`Next` can advance between pulling the `Ready` entries and the range
controller handling them, leading to an assertion failure. Downgrade the
assertion to a warning and introduce the same assertion when pulling
entries.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: decrease logging verbosity

Drop more expensive VInfos from 1 to 2 or 3.

Epic: none
Release note: None

kvoli committed:
kvserver: swallow raft no store handler errors when admitted > 0

Similar to v1, we annotate raft request batches with an additional
request, which includes the admitted entry indexes for a range.

Don't try to handle these messages after processing the admitted
indices.

Epic: none
Release note: None

kvoli committed:
kvserver: downgrade next != last entry idx to warning and relocate

...test which doesn't transfer the lease. The lease transfer test is
reliably hitting the [-1] panic on marshaling the `RaftMessageRequest`
struct, along with the relocate with lease transfer.

The added test, relocate without a lease transfer, does not hit this
problem.

Epic: none
Release note: None

kvoli committed:
raft: ensure admitted marks are copied, instead of referenced

Fixes -1 out of bounds marshaling panic.

Epic: none
Release note: None

kvoli committed:
.*: fix linter errors and copy admitted entries

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: fix invariant log statement

Epic: none
Release note: None

kvoli committed:
kvserver: add 3 node restart test to rac_v2 integration tests

1. writes some data
2. stops n3
3. writes more data
4. starts n3
5. writes more data

Hoping to repro index_to_send > queue.index_to_send assertion panic from
roachprod.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: fix snap->replicate transition

Would cause `(arg)index_to_send=23812 > queue.index_to_send=15079 ` to
fire as probeToReplicate was incorrectly called, instead of
snapshotToReplicate.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: detail wait for eval assertion

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: lock stream on init

To avoid failing the race detector.

Epic: none
Release note: None

sumeerbhola committed:
handle leader not in set of replicas

sumeerbhola committed:
fix incorrect assertion in WaitForEval

kvoli committed:
kvserver: synchronize with repl.mu

There were various places where it was assumed that replica.mu would be
held if mutating raft state. This wasn't true for rac v2. Update the
raft interface and range controller methods to explicitly require
replica.mu to be RLocked or Locked and reflect this in updated method
naming.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: add nolint deferunlockcheck

For testing only.

Epic: none
Release note: None

kvoli committed:
kvserver: add back last entry idx

`LastEntryIndex` is used in debugging. With the locking changes in the
prior commit, calling LastEntryIndex would deadlock. Differentiate
whether the replica send stream is being initialized for the leader or
follower, when its a leader we know the replica mutex must already be
held; when its for a follower, it is not.

Epic: none
Release note: None

sumeerbhola committed:
call makeConsistentInStateReplicateLocked in more cases to reduce code duplication

kvoli committed:
kvserver: fix repl.mu deadlock

LastEntryIndex would not unlock the replica mu after read locking. Defer
the unlock. Also clean up the replica mu locking in
flow_control_replica_integration handleRaftEvent.

Epic: none
Release note: None

sumeerbhola committed:
fix makeConsistentWhenUnexpectedPopLocked: we need to fix stats in all cases
since send-queue has been popped from

kvoli committed:
.*: add rac v2 metrics

Epic: none
Release note: None

kvoli committed:
raft: wrap entering state snapshot error

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: add info to inherited wc assertion

Also include the state of the send stream and replica state info.

Epic: none
Release note: None

sumeerbhola committed:
remove inheritedPri assertion and the inheritance logic.
Since we only consume 4MB of tokens when scheduled, there may be
remaining tokens of regular work class, but we may have already
popped all the regular work class entries. In this case we lower
bound the inherited priority to be RaftNormalPri instead of
incorrectly fataling.

sumeerbhola committed:
makeConsistentWhenUnexpectedPopLocked needs to return all eval tokens
since the entries popped have unknown size and are not being tracked.
If we don't do this, we will leak eval tokens.

kvoli committed:
base: disable raft quota and max inflights

Disable the quota pool and maximum inflight messages/bytes by setting a
large value. Replication flow control now performs this function.

NOTE: This is a hack, we should deprecate and integrate RAC into these
functions explicitly.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: don't include tracker in stream string

The `Tracker.String()` method includes every tracked entry, which can be
in the thousands+. Print the stream instead e.g. `t1/s1`.

Epic: none
Release note: None

kvoli committed:
roachtest: ignore latency below 10ms in perturbation tests

The admission control perturbation latency tests assert that the pXX
latency does not increase by more than some % when perturbing, and post
perturbation, when compared to the baseline before perturbation.

When the latency being measured is small, the relative increases can be
much larger, failing the test, despite a similar user quality of
service.

Ignore latency results below 10ms when checking if a change is
acceptable. i.e. latency results less than 10ms are considered
acceptable regardless of the baseline.

Epic: CRDB-37514
Release note: None

pav-kv committed:
raft: fix term race in SetAdmitted

Send a MsgAppResp only if the admitted indices are in the context of a
new leader, i.e. the last accepted term == raft.Term.

Epic: none
Release note: none

kvoli committed:
kvflowconnectedstream: copy scheduled map

Avoid concurrently iterating and deleting entries in the scheduled
replica map.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: sync scheduled replicas

Epic: none
Release note: None

kvoli committed:
roachtest: make ac latency output order consistent

The final validation output was not consistent as it relied on iterating
over a map.

Sort the map keys first, to allow a consistent output order

Epic: none
Release note: None

kvoli committed:
kvserver: inform lease transfer with flow state

Previously, it was possible to transfer a lease to a replica which was
behind or likely struggle keeping once becoming the leaseholder.

Introduce a new `RangeController` method `ValidLeaseTarget`, which
determines if the given replica would be a suitable leaseholder, based
off the state of its send queue, eval/send tokens and stream state.

Epic: none
Release note: None

kvoli committed:
roachtest: extend dead store timer in ac latency test

Avoid stores up-replicating away from the target node, reducing the
backlog of work.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: include tokens deducted in blocked stream log

Include the delta (returned-deducted) as well as the deducted tokens
when logging streams which are blocked.

Epic: none
Release note: None

sumeerbhola committed:
hack42: send queue is always drained using RaftLowPri

There is some justification in a code comment.

kvoli committed:
kvflowconnectedstream: update blocked stream log spacing

For readability, include spaces between '-' of returned and deducted
stream tokens.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: include stream <eval/send> in blocked stream log

To more easily determine which stream type is blocked (no tokens).

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: pprint negative token values

When printing token values, they were first converted to an unsigned
integer, which causes erroneous output. Introduce a pretty print
function for tokens, which accounts for negative tokens and formats
correctly.

Epic: none
Release note: None

kvoli committed:
kvserver: limit needs lease queue spam

We would previously log on every occurrence of a non-leaseholder replica
not processing through the replica queue, due to requiring a lease. This
occurs semi-frequently in a dynamic cluster, with leaseholder changes.

Mute needs lease logging, other queue errors will still be logged.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: comment unreachable hack42 code

Required to pass linter to build.

Epic: none
Release note: None

kvoli committed:
kvflowconnectedstream: remove deadlock on token setting change

Remove re-entrant deadlock within the setting change handler for v2
tokens.

Epic: none
Release note: None

sumeerbhola committed:
add an interface to aid in understanding integration calls

sumeerbhola committed:
clarify how to fix a todo regarding wasteful ready enqueueing

sumeerbhola commented:
RACv2 is fully implemented. Closing this prototype.
